gpytorch/lazy/interpolated_lazy_variable.py

```diff
@@ -0,0 +1,280 @@
+                left_factor = left_factor.unsqueeze(0)
+                right_factor = right_factor.unsqueeze(0)
+
+            # Left factor
+            left_interp = _make_sparse_from_indices_and_values(left_interp_indices,
+                                                               left_interp_values,
+                                                               self.base_lazy_variable.size()[-2])
+            left_factor = left_factor.transpose(-1, -2)
+            left_res = bdsmm(left_interp, left_factor).transpose(-1, -2).contiguous()
+
+            # Right factor
+            right_interp = _make_sparse_from_indices_and_values(right_interp_indices,
+                                                                right_interp_values,
+                                                                self.base_lazy_variable.size()[-1])
+            right_factor = right_factor.transpose(-1, -2)
+            right_res = bdsmm(right_interp, right_factor).transpose(-1, -2).contiguous()
+
+            res = tuple(list(base_lazy_var_deriv(left_res, right_res)) + [None] * 4)
+            return res
+
+        return closure
+
+    def _transpose_nonbatch(self):
+        res = self.__class__(self.base_lazy_variable.transpose(-1, -2), self.right_interp_indices,
+                             self.right_interp_values,
+                             self.left_interp_indices, self.left_interp_values)
+        return res
+
+    def diag(self):
+        n_data, n_interp = self.left_interp_indices.size()
+
+        # Batch compute the non-zero values of the outer products w_left^k w_right^k^T
+        left_interp_values = self.left_interp_values.unsqueeze(2)
+        right_interp_values = self.right_interp_values.unsqueeze(1)
+        interp_values = torch.matmul(left_interp_values, right_interp_values)
+
+        # Batch compute Toeplitz values that will be non-zero for row k
+        left_interp_indices = self.left_interp_indices.unsqueeze(2).expand(n_data, n_interp, n_interp).contiguous()
+        right_interp_indices = self.right_interp_indices.unsqueeze(1).expand(n_data, n_interp, n_interp).contiguous()
+        base_var_vals = self.base_lazy_variable._get_indices(left_interp_indices.view(-1),
+                                                             right_interp_indices.view(-1))
+        base_var_vals = base_var_vals.view(left_interp_indices.size())
+
+        diag = (interp_values * base_var_vals).sum(2).sum(1)
+        return diag
+
+    def mul(self, other):
+        """
+        Multiplies the matrix by a constant, or elementwise the matrix by another matrix
+        """
+        if not (isinstance(other, Variable) or isinstance(other, LazyVariable)) or \
+               (isinstance(other, Variable) and other.numel() == 1):
+            base_lazy_variable = ConstantMulLazyVariable(self.base_lazy_variable, other)
+
+            return self.__class__(base_lazy_variable, self.left_interp_indices, self.left_interp_values,
```

Comment:
Does it matter whether we do `W(cT)W'` vs `cWTW'`? I think it would be cleaner to just call `super` here and use `LazyVariable`'s `mul`.

Reply:
The reason for this has to do with the new GridInducingPointModules, but I think after I make some changes to GridInducingPointModule it won't matter.